src/KDSoapClient/KDSoapUdpClient.cpp

```diff
@@ -0,0 +1,111 @@
+#include <QNetworkDatagram>
+#include <QNetworkInterface>
+
+KDSoapUdpClient::KDSoapUdpClient(QObject *parent) : 
+    QObject(parent),
+    d_ptr(new KDSoapUdpClientPrivate(this))
+{
+    Q_D(KDSoapUdpClient);
+    d->socketIPv4 = new QUdpSocket(this);
+    connect(d->socketIPv4, &QUdpSocket::readyRead, [d]() { d->readyReadIPv4(); });
+    d->socketIPv6 = new QUdpSocket(this);
+    connect(d->socketIPv6, &QUdpSocket::readyRead, [d]() { d->readyReadIPv6(); });
+}
+
+bool KDSoapUdpClient::bind(quint16 port) {
+    Q_D(KDSoapUdpClient);
+    bool rc = true;
+    rc = d->socketIPv4->bind(QHostAddress::AnyIPv4, port, QUdpSocket::ShareAddress) && rc;
+    rc = d->socketIPv6->bind(QHostAddress::AnyIPv6, port, QUdpSocket::ShareAddress) && rc;
+    return rc;
+}
+
+bool KDSoapUdpClient::sendMessage(const KDSoapMessage &message, const KDSoapHeaders &headers, const QHostAddress& address, quint16 port) {
+    Q_D(KDSoapUdpClient);
+    KDSoapMessageWriter msgWriter;
+    msgWriter.setVersion(KDSoap::SOAP1_2);
+    const QByteArray data = msgWriter.messageToXml(message, QString(), headers, QMap<QString, KDSoapMessage>());
+    
+    if(address.isMulticast()) {
+        bool anySuccess = false;
+        for (const auto &iface : QNetworkInterface::allInterfaces()) {
+            if(iface.flags().testFlag(QNetworkInterface::IsUp) &&
+                iface.flags().testFlag(QNetworkInterface::CanMulticast)) {
+                //qDebug() << "Sending multicast to" << iface.name() << address << ":" << data;
+                if(address.protocol() == QAbstractSocket::IPv4Protocol) {
+                    d->socketIPv4->setMulticastInterface(iface);
+                    qint64 writtenSize = d->socketIPv4->writeDatagram(QNetworkDatagram(data, address, port));
+                    anySuccess = (writtenSize == data.size()) || anySuccess;
+                } else if(address.protocol() == QAbstractSocket::IPv6Protocol) {
+                    d->socketIPv6->setMulticastInterface(iface);
+                    qint64 writtenSize = d->socketIPv6->writeDatagram(QNetworkDatagram(data, address, port));
+                    anySuccess = (writtenSize == data.size()) || anySuccess;
+                }
+            }
+        }
+        return anySuccess;
+    } else {
+        //qDebug() << "Sending to" << address << ":" << data;
+        if(address.protocol() == QAbstractSocket::IPv4Protocol) {
+            qint64 writtenSize = d->socketIPv4->writeDatagram(QNetworkDatagram(data, address, port));
+            return writtenSize == data.size();
+        } else if(address.protocol() == QAbstractSocket::IPv6Protocol) {
+            qint64 writtenSize = d->socketIPv6->writeDatagram(QNetworkDatagram(data, address, port));
+            return writtenSize == data.size();
+        }
+    }
+    return false;
+}
+
+void KDSoapUdpClientPrivate::readyReadIPv4()
+{
+    while (socketIPv4->hasPendingDatagrams()) {
+        QNetworkDatagram datagram = socketIPv4->receiveDatagram();
```

Comment:
The two lines could be merged.
In fact, considering this and ipv6 are the same code except for the socket you could just pass the socket in, or cast sender().

Reply:
I changed that code, also for Qt4 compatibility.